core/src/main/java/org/apache/iceberg/BaseSnapshot.java

```diff
@@ -120,14 +122,42 @@ public String operation() {
     return summary;
   }
 
-  @Override
-  public List<ManifestFile> manifests() {
-    if (manifests == null) {
+  private void cacheManifests() {
```

Comment:
Do we have to reason about thread safety like we do in `PartitionSpec`, for example?

Reply:
No. The main problem in #219 was that the lazy value was set before it was completely initialized. Here, `allManifests` is set to a list that will not change.

In that PR, we also added double checking and a lock to prevent two threads from building the same value. We could do that here as well, but it doesn't affect correctness. There is just a race condition to produce the same sets of manifests.

Reply:
I think it would make sense to prevent the race condition later on. I'll submit a PR.